hawkular-inventory-api/src/main/java/org/hawkular/inventory/api/model/AbstractHashTree.java

```diff
@@ -0,0 +1,283 @@
+            this.tctor = tctor;
+            this.cctor = cctor;
+        }
+
+        public This withPath(RelativePath path) {
+            this.path = path;
+            return castThis();
+        }
+
+        public This withHash(H hash) {
+            this.hash = hash;
+            return castThis();
+        }
+
+        public H getHash() {
+            return hash;
+        }
+
+        public RelativePath getPath() {
+            return path;
+        }
+
+        public boolean hasChildren() {
+            return children != null && !children.isEmpty();
+        }
+
+        public void addChild(T childTree) {
+            getChildren().put(childTree.getPath().getSegment(), childTree);
+        }
+
+        public Child startChild() {
+            return cctor.apply(tctor, castThis());
+        }
+
+        protected T build() {
+            if ((path == null || hash == null) && (children != null && !children.isEmpty())) {
+                throw new IllegalStateException("Cannot construct a tree hash node without a path or " +
+                        "hash and with children. While empty tree without a hash or path is OK, having children " +
+                        "assumes the parent to be fully established.");
+            }
+
+            if (path != null) {
+                int myDepth = path.getDepth();
+
+                for (T child : getChildren().values()) {
+                    int childDepth = child.getPath().getDepth();
+                    if (!path.isParentOf(child.getPath()) || childDepth != myDepth + 1) {
+                        throw new IllegalStateException(
+                                "When building a tree node with path " + path + " an attempt " +
+                                        "to add a child on path " + child.getPath() +
+                                        " was made. The child's path must extend the parent's path by exactly" +
+                                        " one segment, which is not true in this case.");
+                    }
+                }
+            }
+
+            return tctor.construct(path, hash, Collections.unmodifiableMap(getChildren()));
+        }
+
+        private Map<Path.Segment, T> getChildren() {
+            if (children == null) {
+                children = new HashMap<>();
+            }
+
+            return children;
+        }
+
+        @SuppressWarnings("unchecked")
+        private This castThis() {
+            return (This) this;
+        }
+    }
+
+    abstract static class AbstractChildBuilder<
+            This extends AbstractChildBuilder<This, Parent, Child, T, H>,
+            Parent extends Builder<?, This, T, H>,
+            Child extends AbstractChildBuilder<Child, This, ?, T, H>,
+            T extends AbstractHashTree<T, H>,
+            H extends Serializable>
+
+            extends AbstractBuilder<This, Child, T, H>
+            implements ChildBuilder<This, Parent, Child, T, H> {
```

Comment:
I forgot there is a AbstractMemeFactoryBuilder on the internets :]
https://memegenerator.net/instance/70760251